classes/userGroup/DAO.php

```diff
@@ -0,0 +1,173 @@
+use Illuminate\Support\LazyCollection;
+use PKP\core\EntityDAO;
+use PKP\services\PKPSchemaService;
+use PKP\facades\Repo;
+use PKP\userGroup\UserGroup;
+use Illuminate\Database\Query\Builder;
+
+class DAO extends EntityDAO
+{
+    /** @copydoc EntityDAO::$schema */
+    public $schema = PKPSchemaService::SCHEMA_USER_GROUP;
+
+    /** @copydoc EntityDAO::$table */
+    public $table = 'user_groups';
+
+    /** @copydoc EntityDAO::$settingsTable */
+    public $settingsTable = 'user_group_settings';
+
+    /** @copydoc EntityDAO::$primarykeyColumn */
+    public $primaryKeyColumn = 'user_group_id';
+
+    /** @copydoc EntityDAO::$primaryTableColumns */
+    public $primaryTableColumns = [
+        'id' => 'user_group_id',
+        'contextId' => 'context_id',
+        'roleId' => 'role_id',
+        'isDefault' => 'is_default',
+        'showTitle' => 'show_title',
+        'permitSelfRegistration' => 'permit_self_registration',
+        'permitMetadataEdit' => 'permit_metadata_edit',
+    ];
+
+    /**
+     * Instantiate a new DataObject
+     */
+    public function newDataObject(): UserGroup
+    {
+        return app(UserGroup::class);
+    }
+
+    /**
+     * Get the total count of rows matching the configured query
+     */
+    public function getCount(Collector $query): int
+    {
+        return $query
+            ->getQueryBuilder()
+            ->count();
+    }
+
+    /**
+     * Get a list of ids matching the configured query
+     */
+    public function getIds(Collector $query): Collection
+    {
+        return $query
+            ->getQueryBuilder()
+            ->select('a.' . $this->primaryKeyColumn)
+            ->pluck('a.' . $this->primaryKeyColumn);
+    }
+
+    /**
+     * Get a collection of publications matching the configured query
+     */
+    public function getMany(Collector $query): LazyCollection
+    {
+        $rows = $query
+            ->getQueryBuilder()
+            ->get();
+
+        return LazyCollection::make(function () use ($rows) {
+            foreach ($rows as $row) {
+                yield $row->user_group_id => $this->fromRow($row);
+            }
+        });
+    }
+
+    /**
+     * @copydoc EntityDAO::fromRow()
+     */
+    public function fromRow(object $row): UserGroup
+    {
+        $userGroup = parent::fromRow($row);
+
+        return $userGroup;
+    }
+
+    /**
+     * @copydoc EntityDAO::insert()
+     */
+    public function insert(UserGroup $userGroup): int
+    {
+        return parent::_insert($userGroup);
+    }
+
+    /**
+     * @copydoc EntityDAO::update()
+     */
+    public function update(UserGroup $userGroup)
+    {
+        parent::_update($userGroup);
+    }
+
+    /**
+     * @copydoc EntityDAO::delete()
+     */
+    public function delete(UserGroup $userGroup)
+    {
+        parent::_delete($userGroup);
+    }
+
+    /**
+     * @copydoc EntityDAO::deleteById()
+     */
+    public function deleteById(int $userGroupId)
+    {
+        DB::beginTransaction();
```

Comment:
I'd recommend setting `user_user_groups` and `user_group_stage` to onDelete('cascade')`. That way you could get rid of this entire function.

Reply:
Added the foreignKeys 

```
        Schema::table('user_group_settings', function (Blueprint $table) {
            $table->foreign('user_group_id')->references('user_group_id')->on('user_groups')->onDelete('cascade');
        });

        Schema::table('user_user_groups', function (Blueprint $table) {
            $table->foreign('user_group_id')->references('user_group_id')->on('user_groups')->onDelete('cascade');
        });

        Schema::table('user_group_stage', function (Blueprint $table) {
            $table->foreign('user_group_id')->references('user_group_id')->on('user_groups')->onDelete('cascade');
        });
```
to 
- Upgrade Migration (up, down)
- Installation

Removed the `deleteById` function